XpressAI/xircuits
# ✨ Dynamic Ports

# Description

This PR introduces Dynamic Ports, a specialized category of ports designed to dynamically expand when linked, while still being treated as a singular variable in the compiled code. This feature is useful in workflows that requires the concatenation of multiple unspecified inputs, such as appending chats as lists.

![DynaPorts](https://github.com/XpressAI/xircuits/assets/68586800/47c39215-6874-430c-b4c1-cad6c848818e)


As a general guideline, the behaviors listed below are inherent to dynaports:

- `DynaList`, `DynaTuple`, and `DynaDict` are processed as normal lists, tuples, and dicts in the compiled code, with the order preserved according to the dynaport sequence.
- When a link is created to an empty dynaport, a new dynaport with an identical variable name is spawned below it. A `[x]` label is updated for every subsequently linked port.
- All types of `Literal`s can be linked to DynaLists and DynaTuples, but only `Literal Dicts` can be linked to DynaDicts.
- When a link to a dynaport is removed, the following dynaports are automatically updated.
- If a link is created to a dynaport that already has a link, the new link assumes the current dynaport, and the existing link, along with the following dynaports, are shifted.
- Dynaport inherits other default port behaviors, such as reloading.

Internally, this PR also introduces several essential typings for port models, including `dataType` and `varName`. Previously, these were inferred by performing regex operations on a `model.name`. Now, they are explicitly typed as such.

## References


## Pull Request Type

- [ ] Others (Specify if any)

## Type of Change


# Tests

### 1. Testing DynaPorts
    1. Drag in DynaPorts from XAI_Template.
    2. Verify each `DynaList`, `DynaTuple`, and `DynaDict` have special labels.

### 2. Testing Linking of DynaPorts
    1. Link to an empty DynaPort.
    2. Verify a new dynamic port spawns right below the linked port.
    3. Verify the new dynamic port displays the correct port position [1].
    4. Verify ability to link all types of `Literal`s to DynaLists and DynaTuples, and `Literal Dicts` to DynaDicts.

### 3. Testing Compilation of DynaPorts
    1. Compile code and verify DynaList, DynaTuple, and DynaDict are correctly parsed in their data types and the sequence is accurate.

### 4. Testing Shifting of DynaPorts
    1. Connect a link to a port that already has a link.
    2. Confirm that all subsequent ports are correctly shifted.

### 5. Testing Deletion of DynaPorts
   1. For Single Link
        1. Delete the link and verify the links are properly shifted.
        2. If it's the last port connected, confirm the label [1] disappears and only the main label is visible.
   2. For Multiple Entities
        1. Delete multiple links connected to dynaports and verify they correctly shift.
   3. For Nodes Connected to DynaPorts
        1. Delete nodes (e.g., parameter nodes) that are connected to dynaports without selecting the links themselves and verify they shift correctly.

### 6. Testing Reloading of DynaPorts
    1. Reload a component that has dynamic port links.
    2. Confirm all links are intact without any errors.

## Tested on?

- [ ] Others  (Specify if any)

## Notes

For old workflows, please run the `Reload all nodes` button to reload your nodes so it updates with the latest format.
